torch/distributed/_functional_collectives.py

```diff
@@ -99,43 +98,39 @@ def _wait_tensor(tensor: torch.Tensor) -> torch.Tensor:
         _wait_and_clear_tensor(data_ptr, version_and_work[0])
     return tensor
 
-
 class AsyncCollectiveTensor(torch.Tensor):
-    r"""
-    A Tensor subclass that is only used in eager mode, to hold a 'work' object
-    and then wait on it before invoking a real op.
-
-    Usage, from inside functional collective:
-    def functional_collective(input):
-        input = input.clone()
-        mutated_input, work = c10d.{inplace_collective}(input)
-        return AsyncCollectiveTensor(mutated_input, work)
-    """
-    _tensor: torch.Tensor
+    elem: torch.Tensor
+
+    __slots__ = ['elem']
 
-    __torch_function__ = _disabled_torch_function_impl
+    __torch_function__ = torch._C._disabled_torch_function_impl
 
     @staticmethod
-    def __new__(cls, tensor: torch.Tensor):
-        t = tensor
-        r = torch.Tensor._make_subclass(cls, t, require_grad=t.requires_grad)
-        r._tensor = tensor  # type: ignore[attr-defined]
+    def __new__(cls, elem: torch.Tensor):
+
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls, elem.size(),
+            strides=elem.stride(), storage_offset=elem.storage_offset(),
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=False
```

Comment:
should the requires_grad be the same as `elem` too? or we want the wrapper class to not recording the gradient?

Reply:
Grads and wrappers is confusing.
We want only one of the two to support grad and since we don't rewrap the tensor, we should track grads through the wrapped tensor and not the wrapper.